lib/apiservers/service/restapi/handlers/common.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/cmd/vic-machine/common"
+	"github.com/vmware/vic/lib/apiservers/service/restapi/handlers/util"
+	"github.com/vmware/vic/lib/install/data"
+	"github.com/vmware/vic/lib/install/validate"
+	"github.com/vmware/vic/pkg/version"
+	"github.com/vmware/vic/pkg/vsphere/vm"
+)
+
+func buildData(ctx context.Context, url url.URL, user string, pass string, thumbprint *string, datacenter *string, computeResource *string) (*data.Data, error) {
+	d := data.Data{
+		Target: &common.Target{
+			URL:      &url,
+			User:     user,
+			Password: &pass,
+		},
+	}
+
+	if thumbprint != nil {
+		d.Thumbprint = *thumbprint
+	}
+
+	if datacenter != nil {
```

Comment:
hmm... is this necessary? saw this line: https://github.com/vmware/vic/pull/6318/files#diff-adbd7d5a586ec7aeba3724c39fc6a462R78

Reply:
This block translates a datacenter id (what we want the API to use) into a datacenter name (what the rest of the product uses). The `validateTarget` method is called here with an incomplete data object and the `validator` which is returned is only used within this block. (We can't look up the datacenter name if we have an invalid target or credentials.)

In all cases, including when `datacenter` is `nil`, `validateTarget` is called later to validate the full data object returned by this method.